### PR TITLE
Add penalty term to the two_stream PDE

### PR DESCRIPTION
**Pull request by ckendrick**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Adds the penalty term to the `two_stream` PDE. This also fixes an issue with the coefficients for the penalty term and an issue in the IMEX time advance where the maximum energy returned was not returning the absolute value.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [x] Bugfix
- [x] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/solver.cpp [hunk 1]
```diff
@@ -25,19 +26,19 @@ P simple_gmres(fk::matrix<P> const &A, fk::vector<P> &x, fk::vector<P> const &b,
...
+             fk::vector<P> &x, fk::vector<P> const &b, fk::matrix<P> const &M,
+             int const restart, int const max_iter, P const tolerance,
+             imex_flag const imex, P const alpha_in)
 {
-  auto euler_operator = [&pde, &elem_table, &program_options, &my_subgrid,
-                         imex](fk::vector<P> const &x_in, fk::vector<P> &y,
-                               P const alpha = 1.0, P const beta = 0.0) {
+  auto euler_operator = [&pde, &elem_table, &program_options, &my_subgrid, imex,
+                         alpha_in](fk::vector<P> const &x_in, fk::vector<P> &y,
+                                   P const alpha = 1.0, P const beta = 0.0) {
```
Comment (quantumsteve) [inline src/solver.cpp]: What is the difference between `alpha_in` and `alpha`?

File: src/coefficients.cpp [hunk 2]
```diff
@@ -353,11 +355,13 @@ fk::matrix<P> generate_coefficients(
                 (-1);
           }
           trace_value_3 =
-              (legendre_poly_R_t * legendre_poly_R) * (+1 * flux_right / 2) +
+              (legendre_poly_R_t * legendre_poly_R) * (+1 * flux_right / 2) *
```
Comment (quantumsteve) [inline src/coefficients.cpp:358]: Is `flux-right` type `P`? Mixing integer and floating point arithmetic is risky. 
```suggestion
              (legendre_poly_R_t * legendre_poly_R) * (0.5 * flux_right) *
```
This change may repeated enough times in this file to justify a second PR.

File: src/solver.cpp (hunk 1 above)
Comment (ckendrick) [reply src/solver.cpp]: These changes came from #511 which this PR is based off of. `alpha_in` is a multiplier on the `dt` which is needed in the [second implicit update step of the IMEX advance](https://github.com/project-asgard/asgard/blob/78b11c0a0ac47d053331d2d57048a3086b56ba0d/src/time_advance.cpp#L686).

File: src/pde/pde_collisional_landau.hpp [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+  {
+    ignore(x);
+    ignore(time);
+    return -1.0;
+  }
+
+  static P e1_g2(P const x, P const time = 0)
+  {
+    ignore(time);
+    return (x > 0.0) ? x : 0.0;
```
Comment (quantumsteve) [inline src/pde/pde_collisional_landau.hpp]:
```suggestion
    return std::max(P{0.0}, x);
```

File: src/pde/pde_collisional_landau.hpp [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+  static P get_dt_(dimension<P> const &dim)
+  {
+    ignore(dim);
+    /* return dx; this will be scaled by CFL from command line */
+    // return std::pow(0.25, dim.get_level());
+
+    // TODO: these are constants since we want dt always based on dim 2,
+    //  but there is no way to force a different dim for this function!
+    // (Lmax - Lmin) / 2 ^ LevX * CFL
+    return (6.0 - (-6.0)) / std::pow(2, 3);
```
Comment (quantumsteve) [inline src/pde/pde_collisional_landau.hpp]:
```suggestion
    return P{1.5};
```

File: src/pde/pde_collisional_landau.hpp [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+  inline static std::vector<vector_func<P>> const exact_vector_funcs_ = {};
+  inline static scalar_func<P> const exact_scalar_func_               = {};
+
+  static P get_dt_(dimension<P> const &dim)
+  {
+    ignore(dim);
+    /* return dx; this will be scaled by CFL from command line */
+    // return std::pow(0.25, dim.get_level());
+
```
Comment (quantumsteve) [inline src/pde/pde_collisional_landau.hpp]:
```suggestion
```

File: src/solver.cpp (hunk 1 above)
Comment (mkstoyanov) [reply src/solver.cpp]: Do we have better name than `alpha`, e.g., penalty/relaxation/mixing coefficient.

File: src/coefficients.cpp (hunk 2 above)
Comment (ckendrick) [reply src/coefficients.cpp:358]: This should probably be pulled into a second PR. This happens all over the file.

File: src/pde/pde_collisional_landau.hpp (hunk 4 above)
Comment (ckendrick) [reply src/pde/pde_collisional_landau.hpp]: I want to leave this as-is so it is more clear where this value came from: (Lmax - Lmin) / 2 ^ Lev
The issue is this needs to be using the values from the second dimension. Currently, the `get_dt` function only passes the first dimension in.

**Comment by mkstoyanov** [comment]
This needs a major rebase, I won't be able to review today.

File: src/solver.hpp [hunk 6]
```diff
@@ -26,7 +26,7 @@ simple_gmres(PDE<P> const &pde, elements::table const &elem_table,
              fk::vector<P> &x, fk::vector<P> const &b, fk::matrix<P> const &M,
              int const restart, int const max_iter, P const tolerance,
              imex_flag const imex = imex_flag::unspecified,
-             const P alpha_in     = 1.0);
+             const P dt_factor    = 1.0);
```
Comment (quantumsteve) [inline src/solver.hpp:29]: Would a test with dt_factor != 1.0 be helpful?

**Review by quantumsteve — APPROVED** [review]
LGTM.

**Pull request merged**